cmd/sts-handlers.go

```diff
@@ -147,6 +145,29 @@ func (sts *stsAPIHandlers) AssumeRole(w http.ResponseWriter, r *http.Request) {
 	ctx = newContext(r, w, action)
 	defer logger.AuditLog(w, r, action, nil)
 
+	subPolicyStr := r.Form.Get("Policy")
+	// https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html
+	// The plain text that you use for both inline and managed session
+	// policies shouldn't exceed 2048 characters.
+	if len(subPolicyStr) > 2048 {
+		writeSTSErrorResponse(w, stsErrCodes.ToSTSErr(ErrSTSInvalidParameterValue))
+		return
+	}
+
+	if len(subPolicyStr) > 0 {
+		subPolicy, err := iampolicy.ParseConfig(bytes.NewReader([]byte(subPolicyStr)))
+		if err != nil {
+			writeSTSErrorResponse(w, stsErrCodes.ToSTSErr(ErrSTSInvalidParameterValue))
+			return
+		}
+
+		// Version in policy must not be empty
+		if subPolicy.Version == "" {
+			writeSTSErrorResponse(w, stsErrCodes.ToSTSErr(ErrSTSInvalidParameterValue))
+			return
+		}
```

Comment:
@y4m4, the subpolicy is not being validated against the main policy for the IAM user. Even  if subpolicy is not valid w.r.t the user policy - assume role API returns a session token and temp credentials. The policies are applied only at the time of client requesting some resource. Seems like there should be a check to validate if subpolicy nests inside main user policy

Reply:
We cannot do that, a policy can only be validated when we know the API and the resource.  Temp creds should be generated regardless of the correct or incorrect policy. 

IT is the application which is responsible for setting the right policies.